All _target=blank links should also have rel='noreferrer noopener'

peterbe opened the pull request:
@davidflanagan r?

I did not know how to test this locally so I'm reaching out for help. 

Also, I'm not sure why I don't have the ability to request a review the regular GitHub way.

peterbe committed:
All _target=blank links should also have rel='noreferrer noopener'

davidflanagan reviewed (approved):
This looks good to me, Peter.

The reason you couldn't request a review is probably that you aren't a member of the "KumaScript Peers" group. @Elchi3 and @jwhitlock are maintainers, so I think that they are the two people who could (and should) add you.

To test this locally, you're first going to have to know of a document that uses this macro. If it isn't in the local database, you could scrape it with the scrape_document management command to get a local copy. Or you could just edit a document on your dev server to add the macro call.

Then, assuming you've done a `docker-compose restart kumascript` I think you ought to be able to test by just doing a Cmd-Shift-R to re-run the macros on the page, and then use view source or devtools to verify that the rel attribute has been added.

Also: feel free to add a real test case to the macro tests. We need all we can get.

Elchi3 commented:
> The reason you couldn't request a review is probably that you aren't a member of the "KumaScript Peers" group. @Elchi3 and @jwhitlock are maintainers, so I think that they are the two people who could (and should) add you.

You are now a KumaScript peer. Welcome!

peterbe commented:
Locally or on prod, how can I find if `mediawiki.external.ejs` is in use? 
I have the Django ORM so I can search for things like `{{Foo}}` on all the revisions' content. But I don't know what "Foo" is. ...based on only knowing the name `mediawiki.external.ejs`.

davidflanagan commented:
I think we might have a dashboard somewhere that shows which pages use which macros. I'm not sure if it works.  IIRC macro names are the same as the filenames and are case insensitive. So I think you'd need to look for mediawiki.external, case insensitive inside double curly braces. There might be arguments inside parens, so probalby only look for {{ followed by optional space followed by mediawiki.external, I think.

escattone commented:
@peterbe @davidflanagan https://developer.mozilla.org/en-US/dashboards/macros provides what you want, and it is up-to-date.

peterbe commented:
@escattone That's awesome! So it's simply called `mediawiki.external`. I had no idea. I thought the `mediawiki` some sort of namespace and thought perhaps it would become `{{External('foo')}}` or something like that. 

In my mysql dump, I don't have any though:

```
>>> Revision.objects.filter(content__contains='mediawiki.external')
<QuerySet []>
```

So either I try to inject it by myself, or I try to look for it in prod so see it used for real.

peterbe commented:
Ok. So I just typed this into one of the wiki pages:
```
{{mediawiki.external('https://www.peterbe.com/')}}
```
It rendered to:
```html
<a class="external" href="" rel="noopener noreferrer" title="">https://www.peterbe.com/</a>
```

Note, no `target="_blank"` and no `href` or `title` value!

Perhaps I just don't get how it works. Also, perhaps I'm making a bigger deal of this than necessary but by stirring here I think I'm learning. 

[This page](https://developer.mozilla.org/en-US/docs/Mozilla/Projects/NSS/Reference/FC_DestroyObject$edit) for example uses `mediawiki.external` but as the only parameter `in` which doesn't make much sense rendered. 

And on https://developer.mozilla.org/en-US/docs/Mozilla/Projects/Rhino/Shell$edit it uses it like this: 

```

<p><code>java org.mozilla.javascript.tools.shell.Main {{ mediawiki.external('<var>options</var>') }} <var>script-filename-or-url</var> {{ mediawiki.external('<var>script-arguments</var>') }} </code></p>
```
which produces:
```html
<p><code>java org.mozilla.javascript.tools.shell.Main [<var>options</var>] <var>script-filename-or-url</var> [<var>script-arguments</var>] </code></p>
```

So I see no good sensible way to make it produce something like:
```html
<a class="external" href="https://www.peterbe.com/" target="_blank" rel="noopener noreferrer" title="https://www.peterbe.com/">https://www.peterbe.com/</a>
```
which would kinda make sense.

ExE-Boss commented:
This is because `target="_blank"` is automatically removed by the **Kuma** HTML sanitiser.

ExE-Boss commented on macros/mediawiki.external.ejs:
```diff
@@ -14,7 +14,7 @@ if (/^(http[s]?:\/\/|mailto:)/.test('' + $0)) {
   }
 
-  str = '<a class="external" href="' + link + '" title="' + link + '" target="_blank">' + text + '</a>';
+  str = '<a class="external" href="' + link + '" title="' + link + '" target="_blank" rel="noopener noreferrer">' + text + '</a>';
```
```suggestion
  str = '<a class="external" href="' + link + '" title="' + link + '>' + text + '</a>';
```

peterbe replied:
I like this, but first I'd like to understand this crazy macro :)

peterbe replied:
Sorry @ExE-Boss but I've forgotten the justification for your suggestion. Why remove the `target="_blank"`? Because if we remove it, what's the point of bothering with the `rel="noopener noreferrer"`?

ExE-Boss reviewed (commented):
This is because `target="_blank"` is automatically removed by the **Kuma** HTML sanitiser.

peterbe commented:
> This is because `target="_blank"` is automatically removed by the Kuma HTML sanitiser.

That is still baffling. Why is the macro called "mediawiki.external" if you're not allowed to use `target="_blank"` (Note: I'm NOT a fan of this attribute and think it should mostly be eradicated except when used in SPAs like Gmail or GitHub)

@ExE-Boss Do you know why the produced HTML becomes:
```html
<a class="external" href="" rel="noopener noreferrer" title="">https://www.peterbe.com/</a>
```
which, to me, seems pretty useless.

ExE-Boss commented:
`{{mediawiki.external}}` is an old macro that was created to handle MediaWiki links when MDN migrated to DekiWiki.

From the [KumaScript macro massacre spreadsheet](https://docs.google.com/spreadsheets/d/1CaNZoKB3fx2tDMvJ_KTke4KS4KFpVBrHpdAisRVjbIk/edit?disco=AAAACTWBXLc):
> This is a left‑over from the migration to DekiWiki from MediaWiki (circa 2008).
>
> It’s mostly present in pages which weren’t really updated since the MediaWiki days.
>
> It should be removed from pages and deleted.

peterbe commented:
It still belongs to some [428 pages](https://developer.mozilla.org/en-US/dashboards/macros) :(

Also, the note is weird: "It should be removed from pages and deleted."
Removed AND deleted? Delete the macro or delete the whole page?!

I'm tempting to take your suggestion for now @ExE-Boss which is to just delete all this `target="_blank"` nonsense and then the `rel` tag is a non-issue.

ExE-Boss commented:
I meant delete the macro.

peterbe commented:
If you do that, what's going to happen to the 428 pages that "depend" on it?

ExE-Boss commented:
We first de‑macroify the calls, like what I did with:
- https://developer.mozilla.org/en-US/docs/Mozilla/Tech/XUL/Tutorial/XBL_Example$compare?locale=en-US&to=1437695&from=1357315#difflib_chg_to66__7
- https://developer.mozilla.org/en-US/docs/Mozilla/Tech/XUL/Introduction_to_XUL$compare?locale=en-US&to=1437714&from=1354453#difflib_chg_to97__2

peterbe commented:
Not sure what that means. That looks like a manual change. 427 left to go?
